tests/Test/Assert/AssertTokensTrait.php

```diff
@@ -33,6 +33,11 @@ private static function assertTokens(Tokens $expectedTokens, Tokens $inputTokens
 
             $inputToken = $inputTokens[$index];
 
+            if (str_contains($inputToken->getContent(), '|')) {
```

Comment:
TODO: to be removed before merging this PR

Reply:
The purpose of this assertion is to ensure that 2 token collections are equal. We need this to be sure that whatever way the collection was generated, the collection generated for same input piece of code is always exactly the same. If it's not, sth is wrong and next fixer, when reusing the collection after previous fixer, may face issues and not be able to adjust the code properly

Reply:
So I'm still no sure what am I suppose to do with it. :)

Reply:
FYI: without this condition there are error like this:

```
1) PhpCsFixer\Tests\Fixer\FunctionNotation\PhpdocToParamTypeFixerTest::testFixPhp80 with data set "union types" ('<?php /** @param Foo|Bar $bar...ar) {}', '<?php /** @param Foo|Bar $bar...ar) {}')
The token at index 7 must be:
{
    "id": 262,
    "name": "T_STRING",
    "content": "Foo",
    "isArray": true,
    "changed": false
},
got:
{
    "id": 262,
    "name": "T_STRING",
    "content": "Foo|Bar",
    "isArray": true,
    "changed": false
}.
Failed asserting that false is true.

```

Reply:
The problem is that `$inputTokens` contains union types in one token (like `Foo|Bar`), while `$expectedTokens` contains multiple tokens (`Foo`, `|` and `Bar`). @keradus what would be the best way to approach it?

Reply:
no, we need to make sure that code on tokens level is valid and equal expectations, and not that only generated code string is.

if first one is not happening, then we have a token with invalid content, that will not work properly for next rule executing.

Reply:
issue is that somewhere, we have a tokens for type.
we shall have 3 tokens, `[Foo, |, Bar]`, but instead you created a token instance that is `Foo|Bar`.

look at example expectation tokens collection , especially indices 10,11,12 - https://3v4l.org/3tJ7P

Reply:
Ok, so we probably need to squash types' tokens into one string token when applying fix for union types. I'll look at it later.

Reply:
not my code

I was replying to MortalFlesh, who asked for next steps - yeah, we have more than me and you in this thread :P

Reply:
> [MortalFlesh](https://github.com/MortalFlesh) on Aug 19, 2022

I believe MortalFlesh forgot previous steps numerous times already, and don't need next ones 🤪.

Reply:
@keradus I've introduced abstract `AbstractPhpdocToTypeDeclarationFixer::createTokensFromRawType()` method for generating `Tokens` from raw union type (simple string). It couldn't be standardised inside `createTypeDeclarationTokens()` because e.g. `static` can be used as return type but can't be used as param type, also index ranges to trim before/after type differ between these fixers that have to use different template for tokenising raw type. I hope it's a good solution here 🙂. I'll close this thread as the condition from trait was removed, if you have doubts about implementation please comment in proper place.